cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/timeseries/FactTable.java

```diff
@@ -121,11 +125,79 @@ public FactScanner scan(FactScan scan) throws Exception {
                              scan.getStartTs(), scan.getEndTs());
   }
 
+  /**
+   * Given a list of previous tags, we return a list of available tags at the child context in the provided time range.
+   * @param previousTags
+   * @param startTs
+   * @param endTs
+   * @return
+   * @throws Exception
+   */
+  public SortedSet<TagValue> getNextTags(List<TagValue> previousTags, long startTs, long endTs) throws Exception {
+    byte[] startRow = codec.createStartRowKey(previousTags, null, startTs, true);
+    byte[] endRow = codec.createEndRowKey(previousTags, null, endTs, true);
+    endRow = Bytes.stopKeyForPrefix(endRow);
+    FuzzyRowFilter fuzzyRowFilter = createFuzzyRowFilter(new FactScan(startTs, endTs, null, previousTags),
+                                                                startRow, true);
+    Row rowResult;
+    int targetIndex = previousTags.size();
+    SortedSet<TagValue> nextLevelContexts = Sets.newTreeSet();
+
+    do {
+      Scanner scanner = null;
+      scanner = timeSeriesTable.scan(startRow, endRow, null, fuzzyRowFilter);
+      rowResult = scanner.next();
+      if (rowResult != null) {
+        //add item to list and increment the start row
+        List<TagValue> tagValues = codec.getTagValues(rowResult.getRow());
+        if (tagValues.size() > targetIndex) {
+          nextLevelContexts.add(tagValues.get(targetIndex));
+        }
+        startRow = codec.getNextRowKey(previousTags, rowResult.getRow());
+      }
+    } while (rowResult != null);
+
+    return nextLevelContexts;
+  }
+
+  /**
+   * Given a list of {@link co.cask.cdap.metrics.store.timeseries.TagValue} and time range,
+   * return the list of metric names with the provided tag value pairs in the given time range.
+   * @param previousTags
+   * @param startTs
+   * @param endTs
+   * @return
+   * @throws Exception
+   */
+  public SortedSet<String> getMeasureNames(List<TagValue> previousTags, long startTs, long endTs) throws Exception {
+    byte[] startRow = codec.createStartRowKey(previousTags, null, startTs, true);
+    byte[] endRow = codec.createEndRowKey(previousTags, null, endTs, true);
+    endRow = Bytes.stopKeyForPrefix(endRow);
+    FuzzyRowFilter fuzzyRowFilter = createFuzzyRowFilter(new FactScan(startTs, endTs, null, previousTags),
+                                                                startRow, true);
+    Row rowResult;
+    SortedSet<String> metricNames = Sets.newTreeSet();
+
+    do {
+      Scanner scanner = null;
+      scanner = timeSeriesTable.scan(startRow, endRow, null, fuzzyRowFilter);
```

Comment:
Add a todo somewhere to push down this logic into server-side of data store, instead of doing it on client-side. Add JIRA for that (don't forget to set "Metrics" as component in it)